checker/src/visitors.rs

```diff
@@ -1508,22 +1497,44 @@ impl<'analysis, 'compilation, 'tcx, E> MirVisitor<'analysis, 'compilation, 'tcx,
         }
     }
 
+    /// Extracts the def_id of a closure, given the type of a value that is known to be a closure.
+    #[logfn_inputs(TRACE)]
+    fn get_def_id_from_closure(&mut self, closure_ty: Ty<'tcx>) -> hir::def_id::DefId {
+        match closure_ty.kind {
+            TyKind::Closure(def_id, _) => {
+                return def_id;
+            }
+            TyKind::Ref(_, ty, _) => {
+                if let TyKind::Closure(def_id, _) = ty.kind {
+                    return def_id;
+                }
+            }
+            _ => {}
+        }
+        self.assume_function_is_angelic = true;
```

Comment:
diag_level == RELAXED?

Reply:
Probably, but I'll come back to this shortly.